### PR TITLE
Refactor api and update drawing logic

### LINKED ISSUE #3: Non-line shapes don't flip over
Shapes other than lines currently have a tendency to grow and shrink drastically, and don't tumble and flip like the line confetti does.

### PR DESCRIPTION
**Pull request by GRA0007**
This is a large PR that will update the major version of this library. Changes the API to return `createConfetti` instead of `showConfetti` from the `useConfetti` hook, and an exported `Canvas` element that requires the user to pass in `canvasRef` from the hook as well.

Also updates the options used to control the confetti. Velocity uses a tuple, and many options have been simplified to use a min/max tuple. Also adds a `wind` option.

Improves the README with option tables.

Fixes #3

### CODE REVIEW

**Review by giraugh — CHANGES_REQUESTED** [review]
Thoughts
- I reckon `...props` instead of exporting a component.
- might be good if the api docs for wind and gravity mentioned that they were horizontal and vertical acceleration
- "range" rather than tuple?
- colors of type `string[]`
- shape functions of type `Record<string, function`
- shape weights of type `Record<string, number>`

File: lib/config/index.js [hunk 1]
```diff
@@ -19,4 +19,11 @@ export const CONFETTI_COLORS = [
   '#607d8b',
 ]
 
-export const TAU = Math.PI * 2
```
Comment (giraugh) [inline lib/config/index.js:22]: :cry:

File: lib/useConfetti.jsx [hunk 2]
```diff
@@ -1,188 +1,172 @@
...
+      const canvasBBox = canvasRef.current.getBoundingClientRect()
+      const { width, height } = canvasBBox
+      canvasBBoxRef.current = canvasBBox
+
+      // Set canvas size
+      const scale = window.devicePixelRatio
+      canvasRef.current.width = Math.floor(width * scale)
+      canvasRef.current.height = Math.floor(height * scale)
+
+      transformMatrix.current = new DOMMatrix().scale(scale)
```
Comment (giraugh) [inline lib/useConfetti.jsx]: This *could* prob be a ReadOnlyDOMMatrix()

File: lib/useConfetti.jsx [hunk 3]
```diff
@@ -1,188 +1,172 @@
...
-      />
-    )
+        if (p.x < -killDistance/window.devicePixelRatio
+          || p.x > (width + killDistance)/window.devicePixelRatio
+          || p.y < -killDistance/window.devicePixelRatio
+          || p.y > (height + killDistance)/window.devicePixelRatio
+        ) return all
+
+        // Render particle
+        const [x, y] = [p.x - canvasBBoxRef.current.left, p.y - canvasBBoxRef.current.top]
```
Comment (giraugh) [inline lib/useConfetti.jsx:147]: I think using `const ctx =` helps with readability and also can be used to do the shorthand object syntax.

File: lib/useConfetti.jsx [hunk 4]
```diff
@@ -1,188 +1,172 @@
...
+        ctxRef.current.setTransform(transformMatrix.current.translate(x, y).rotate(p.angle).scale(1, Math.sin(((p.flip+90)*(Math.PI/180)))))
```
Comment (giraugh) [inline lib/useConfetti.jsx]: Prob worth folding these too for the reduced line length

File: lib/useConfetti.jsx [hunk 5]
```diff
@@ -1,188 +1,172 @@
...
+        shapeFunctions[p.shape]({ p, ctx: ctxRef.current })
+        ctxRef.current.setTransform(1,0,0,1,0,0)
+
+        return [...all, {
+          ...p,
+          vy: p.vy + ((gravity/100) * speed),
+          vx: p.vx + (wind * speed),
+          x: p.x + (p.vx * speed),
+          y: p.y + (p.vy * speed),
+          flip: p.flip + (p.flipIncrement * (rotationVelocityCoefficient ? Math.min(Math.max(p.vy, p.vx), 2) * rotationVelocityCoefficient : 1) * speed),
```
Comment (giraugh) [inline lib/useConfetti.jsx]: I think extracting this line to a variable might also be worth it for readability sake. Like a `// compute rotation amount` block before we do the return

File: lib/util/index.js [hunk 6]
```diff
@@ -1,17 +1,16 @@
 import { useRef, useCallback, useEffect } from 'react'
 
-export const chooseWeighted = (weights, x) => {
-  if (x < 0 || x > 1) throw new Error('x value must be 0-1')
+export const randWeighted = weights => {
   const total = Object.values(weights).reduce((a, b) => a + b, 0)
 
-  let rv = x
+  let rv = Math.random()
```
Comment (giraugh) [inline lib/util/index.js:6]: Might be pref to be `chooseWeighted(weights, randomValue=Math.random())`?

File: package.json [hunk 7]
```diff
@@ -15,7 +17,7 @@
   },
   "scripts": {
     "build": "vite build",
-    "dev:sample": "vite dev",
```
Comment (giraugh) [inline package.json:18]: Still not sold on this change hehe

File: sample/App.jsx [hunk 8]
```diff
@@ -1,62 +1,83 @@
...
+  loadImage,
+  drawImageParticle,
+} from '@stevent-team/react-party'
 
 const App = () => {
   // Setup controls
-  const showConfettiOptions = useControls('ShowConfettiOptions', {
-    duration: { value: 0, min: 0, max: 2000 },
-    count: { value: 75, min: 0, max: 3000 },
+  const launch = useControls('Launch options', {
```
Comment (giraugh) [inline sample/App.jsx]: "Creation options" might be clearer

File: sample/App.jsx [hunk 9]
```diff
@@ -1,62 +1,83 @@
...
-      tiltAngleVariance: { value: 6.28, min: 0, max: 15 },
+
+  const global = useControls('Global options', {
+    gravity: { value: 9.8, min: -50, max: 50, step: .1 },
+    wind: { value: 0, min: -1, max: 1 },
+    speed: { value: 1, min: 0, max: 10 },
+    killDistance: { value: 100, min: 0, max: 3000, step: 10 },
+  }, { collapsed: true })
+
+  const confetti = useControls('Confetti options', {
```
Comment (giraugh) [inline sample/App.jsx]: "Movement options"?

File: sample/App.jsx [hunk 10]
```diff
@@ -1,62 +1,83 @@
...
-    }, { collapsed: true }),
-    tilt: folder({
-      tiltBase: { value: -10, min: -30, max: 30 },
-      tiltVariance: { value: 10, min: -30, max: 30 },
-    }, { collapsed: true }),
-    tiltAngle: folder({
-      tiltAngleBase: { value: 0, min: -30 , max: 30 },
-      tiltAngleVariance: { value: 6.28, min: 0, max: 15 },
+
+  const global = useControls('Global options', {
```
Comment (giraugh) [inline sample/App.jsx]: "Environment Options"?

File: sample/App.jsx [hunk 11]
```diff
@@ -1,62 +1,83 @@
...
+      initialFlip: { value: [0, 360], min: 0, max: 360 },
+      flipIncrement: { value: [-10, 10], min: -50, max: 50 },
     }, { collapsed: true }),
-  })
-  const shapeWeights = useControls('Shape Weights', {
-    'line': { value: 1, min: 0, max: 5, step: 1 },
+    rotationVelocityCoefficient: { value: .8, min: 0, max: 5 },
+  }, { collapsed: true })
+
+  const shapeWeights = useControls('Shape weights', {
```
Comment (giraugh) [inline sample/App.jsx:43]: It might be good to put the default shape weights in a config variable

File: sample/App.jsx [hunk 12]
```diff
@@ -1,62 +1,83 @@
...
+  })
+
+  // Load custom image
+  const imageRef = useRef()
+  useEffect(() => {
+    loadImage("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 36 36'%3E%3Cpath fill='%23DD2E44' d='M35.885 11.833c0-5.45-4.418-9.868-9.867-9.868-3.308 0-6.227 1.633-8.018 4.129-1.791-2.496-4.71-4.129-8.017-4.129-5.45 0-9.868 4.417-9.868 9.868 0 .772.098 1.52.266 2.241C1.751 22.587 11.216 31.568 18 34.034c6.783-2.466 16.249-11.447 17.617-19.959.17-.721.268-1.469.268-2.242z'/%3E%3Cpath fill='%23FDCB58' d='M34.347 23.894l-3.824-1.416-1.416-3.824c-.145-.394-.52-.654-.938-.654-.418 0-.793.26-.938.653l-1.416 3.824-3.824 1.416c-.393.144-.653.519-.653.938 0 .418.261.793.653.938l3.824 1.416 1.416 3.824c.145.393.52.653.938.653.418 0 .793-.261.938-.653l1.416-3.824 3.824-1.416c.392-.145.653-.52.653-.938 0-.418-.261-.793-.653-.937zm-23-16.001l-2.365-.875-.875-2.365C7.961 4.26 7.587 4 7.169 4c-.419 0-.793.26-.938.653l-.876 2.365-2.364.875c-.393.146-.653.52-.653.938 0 .418.26.792.653.938l2.365.875.875 2.365c.146.393.52.653.938.653.418 0 .792-.26.938-.653l.875-2.365 2.365-.875c.393-.146.653-.519.653-.938 0-.418-.26-.792-.653-.938z'/%3E%3C/svg%3E")
```
Comment (giraugh) [inline sample/App.jsx]: Is it possible to load this from a seperate file? Or put it in a constant?

**Review by giraugh — APPROVED** [review]
This is looking great `:D` - I left some thoughts and feedback but its up to you what you change. Also #teamtau

File: lib/config/index.js (hunk 1 above)
Comment (GRA0007) [reply lib/config/index.js:22]: tau is dead, long live tau

**Pull request merged**